File: pipelines/validation/qa.ts
/**
 * Source-data and graph quality assurance for a snapshot.
 *
 *   npm run qa -- <snapshotId>
 *
 * Writes qa-report.json and qa-summary.md into the snapshot directory. Nothing
 * here "repairs" data: the job is to measure and expose, so that a number in
 * the application can be traced to a stated condition of the source.
 */

import { writeFile, readFile } from 'node:fs/promises';
import path from 'node:path';

import { loadSnapshot, snapshotDir } from '../../packages/core/src/snapshot.js';
import { AMDS_MODEL_ASSET_TYPE, AMDS_SURFACE_TYPE } from '../../packages/core/src/types.js';
import { config } from '../lib/config.js';

interface Issue {
  severity: 'error' | 'warning' | 'info';
  issueType: string;
  entityType: 'link' | 'node' | 'graph' | 'restriction';
  count: number;
  detail: string;
  sampleIds?: string[];
}

async function main() {
  const snapshotId = process.argv[2];
  if (!snapshotId) throw new Error('usage: npm run qa -- <snapshotId>');

  const snap = await loadSnapshot(config.dataDir, snapshotId);
  const { graph: g, links, meta } = snap;
  const issues: Issue[] = [];
  const add = (i: Issue) => issues.push(i);

  // ---------------------------------------------------------------- counts
  const byAssetType = new Map<number, number>();
  const bySurface = new Map<number, number>();
  const byOwner = new Map<number, number>();
  const bySpeedSource = new Map<string, number>();
  let oneWay = 0;
  let twoWay = 0;
  let named = 0;
  let lifeline = 0;
  let heavyAllowed = 0;
  let emergencyAllowed = 0;
  let totalLengthM = 0;

  for (const l of links) {
    byAssetType.set(l.modelAssetType ?? -1, (byAssetType.get(l.modelAssetType ?? -1) ?? 0) + 1);
    bySurface.set(l.surfaceType ?? -1, (bySurface.get(l.surfaceType ?? -1) ?? 0) + 1);
    byOwner.set(l.assetOwnerOrganisation ?? -1, (byOwner.get(l.assetOwnerOrganisation ?? -1) ?? 0) + 1);
    bySpeedSource.set(l.speedSource, (bySpeedSource.get(l.speedSource) ?? 0) + 1);
    if (l.oneway === 1) oneWay++;
    else twoWay++;
    if (l.roadName) named++;
    if (l.lifeLineRoute) lifeline++;
    if (l.modeVehicleHeavy) heavyAllowed++;
    if (l.modeEmergency) emergencyAllowed++;
    totalLengthM += l.lengthM;
  }

  // ------------------------------------------------------------ link checks
  const zeroLength = links.filter((l) => !(l.lengthM > 0));
  if (zeroLength.length) {
    add({
      severity: 'error',
      issueType: 'ZERO_LENGTH_LINK',
      entityType: 'link',
      count: zeroLength.length,
      detail: 'Link has non-positive length and cannot carry traffic.',
      sampleIds: zeroLength.slice(0, 10).map((l) => l.amdsId),
    });
  }

  const selfLoops = links.filter((l) => l.sourceNode === l.targetNode);
  if (selfLoops.length) {
    add({
      severity: 'warning',
      issueType: 'SELF_LOOP',
      entityType: 'link',
      count: selfLoops.length,
      detail:
        'Link starts and ends at the same node (typically a closed loop road). ' +
        'Excluded from arc generation; detour results report SOURCE_DATA_ERROR.',
      sampleIds: selfLoops.slice(0, 10).map((l) => l.amdsId),
    });
  }

  const dupAmds = new Map<string, number>();
  for (const l of links) dupAmds.set(l.amdsId, (dupAmds.get(l.amdsId) ?? 0) + 1);
  const dupes = [...dupAmds.entries()].filter(([, c]) => c > 1);
  if (dupes.length) {
    add({
      severity: 'error',
      issueType: 'DUPLICATE_STABLE_ID',
      entityType: 'link',
      count: dupes.length,
      detail: 'Two graph links share an amdsId. Identifiers must be unique within a snapshot.',
      sampleIds: dupes.slice(0, 10).map(([id]) => id),
    });
  }

  const implausible = links.filter((l) => l.lengthM > 50_000);
  if (implausible.length) {
    add({
      severity: 'warning',
      issueType: 'IMPLAUSIBLE_LENGTH',
      entityType: 'link',
      count: implausible.length,
      detail: 'Link longer than 50 km. Verify against the source geometry.',
      sampleIds: implausible.slice(0, 10).map((l) => l.amdsId),
    });
  }

  const flagCounts = new Map<string, number>();
  for (const l of links) for (const f of l.qualityFlags) flagCounts.set(f, (flagCounts.get(f) ?? 0) + 1);

  // ----------------------------------------------------------- graph checks
  const degree = new Int32Array(g.nodeCount);
  for (let a = 0; a < g.arcCount; a++) {
    degree[g.arcFrom[a]]++;
    degree[g.arcTo[a]]++;
  }
  let dangling = 0;
  for (let n = 0; n < g.nodeCount; n++) if (degree[n] <= 2) dangling++;

  const compLinks = new Map<number, number>();
  for (const l of links) {
    if (l.sourceNode < 0) continue;
    const c = g.component[l.sourceNode];
    compLinks.set(c, (compLinks.get(c) ?? 0) + 1);
  }
  const compSizes = [...compLinks.values()].sort((a, b) => b - a);
  const largestShare = compSizes.length ? compSizes[0] / links.length : 0;
  // New Zealand is two main islands with no road connection between them, so a
  // national graph is legitimately dominated by TWO components, not one. Judging
  // it on the largest alone reports Cook Strait as a defect. The pilot extract
  // also spans the strait, so the same rule applies there.
  const topTwoShare =
    compSizes.length > 1 ? (compSizes[0] + compSizes[1]) / links.length : largestShare;

  if (topTwoShare < 0.9) {
    add({
      severity: 'error',
      issueType: 'FRAGMENTED_GRAPH',
      entityType: 'graph',
      count: g.componentCount,
      detail:
        `The two largest connected components hold only ${(topTwoShare * 100).toFixed(1)}% of links ` +
        `(largest ${(largestShare * 100).toFixed(1)}%). A road network should be dominated by ` +
        'one component per landmass. Investigate junction splitting.',
    });
  } else {
    add({
      severity: 'info',
      issueType: 'COMPONENT_STRUCTURE',
      entityType: 'graph',
      count: g.componentCount,
      detail:
        `Largest component ${(largestShare * 100).toFixed(1)}% of links; two largest ` +
        `${(topTwoShare * 100).toFixed(1)}%. Two dominant components is the expected shape ` +
        'for New Zealand: the North and South Islands have no road connection. Smaller ' +
        'components are ferry-only islands (Waiheke, Great Barrier), isolated peninsulas, ' +
        'and off-network parking or access areas.',
    });
  }

  // Closure groups: how many graph links does one physical closure remove?
  const groupSizes = new Map<string, number>();
  for (const l of links) groupSizes.set(l.closureGroupId, (groupSizes.get(l.closureGroupId) ?? 0) + 1);
  const groupDist = new Map<number, number>();
  for (const s of groupSizes.values()) groupDist.set(s, (groupDist.get(s) ?? 0) + 1);

  // Near misses recorded at ingest.
  let nearMisses: any[] = [];
  try {
    nearMisses = JSON.parse(
      await readFile(path.join(snapshotDir(config.dataDir, snapshotId), 'near-misses.json'), 'utf8'),
    );
  } catch {
    /* older snapshots may predate this file */
  }
  if (nearMisses.length) {
    add({
      severity: 'warning',
      issueType: 'UNCONNECTED_NEAR_MISS',
      entityType: 'node',
      count: nearMisses.length,
      detail:
        'A link endpoint lies between 0.05 m and 5 m of another link but was NOT connected. ' +
        'Either the source has a genuine gap or the split tolerance is too tight. ' +
        'These are listed in near-misses.json for review.',
    });
  }

  // Turn restrictions.
  const longRestrictions = snap.restrictions.filter((r) => r.linkSeq.length > 2);
  add({
    severity: longRestrictions.length ? 'warning' : 'info',
    issueType: 'TURN_RESTRICTION_COVERAGE',
    entityType: 'restriction',
    count: snap.restrictions.length,
    detail:
      `${snap.restrictions.length} turn restrictions are applied, of which ` +
      `${longRestrictions.length} span more than two links. Two-link restrictions are ` +
      'enforced exactly by the arc-expanded search; longer sequences are checked against ' +
      'the predecessor chain and are approximate. NOTE: AMDS publishes only 60 restricted ' +
      'turns nationally, so banned-turn coverage is effectively negligible and routes ' +
      'through complex intersections must not be presented as road-legal.',
  });

  // Speed provenance.
  const nslr = bySpeedSource.get('nslr') ?? 0;
  add({
    severity: nslr === 0 ? 'warning' : 'info',
    issueType: 'SPEED_PROVENANCE',
    entityType: 'link',
    count: links.length - nslr,
    detail:
      'AMDS publishes no speed attribute. All time-metric results are derived from ' +
      'estimated speeds and are flagged TIME_ESTIMATED. Distance is the defensible metric.',
  });

  const report = {
    snapshotId,
    generatedAtUtc: new Date().toISOString(),
    snapshot: {
      sourceDataset: meta.sourceDataset,
      retrievedAtUtc: meta.retrievedAtUtc,
      status: meta.status,
      where: meta.where,
      extent: meta.extent,
      analysisExtent: meta.analysisExtent,
      sourceFeatureCount: meta.sourceFeatureCount,
      downloadedFeatureCount: meta.downloadedFeatureCount,
      routableLinkCount: meta.routableLinkCount,
      notes: meta.notes,
    },
    totals: {
      links: links.length,
      arcs: g.arcCount,
      nodes: g.nodeCount,
      components: g.componentCount,
      largestComponentLinks: compSizes[0] ?? 0,
      largestComponentSharePct: Number((largestShare * 100).toFixed(2)),
      twoLargestComponentsSharePct: Number((topTwoShare * 100).toFixed(2)),
      totalNetworkLengthKm: Number((totalLengthM / 1000).toFixed(1)),
      danglingNodes: dangling,
      danglingNodeSharePct: Number(((dangling / g.nodeCount) * 100).toFixed(2)),
      oneWayLinks: oneWay,
      twoWayLinks: twoWay,
      namedLinks: named,
      namedSharePct: Number(((named / links.length) * 100).toFixed(2)),
      lifelineRouteLinks: lifeline,
      heavyVehicleLinks: heavyAllowed,
      emergencyVehicleLinks: emergencyAllowed,
    },
    byModelAssetType: Object.fromEntries(
      [...byAssetType.entries()].map(([k, v]) => [
        (AMDS_MODEL_ASSET_TYPE as any)[k] ?? `code ${k}`,
        v,
      ]),
    ),
    bySurfaceType: Object.fromEntries(
      [...bySurface.entries()].map(([k, v]) => [(AMDS_SURFACE_TYPE as any)[k] ?? `code ${k}`, v]),
    ),
    bySpeedSource: Object.fromEntries(bySpeedSource),
    topOwners: [...byOwner.entries()].sort((a, b) => b[1] - a[1]).slice(0, 15),
    componentSizes: compSizes.slice(0, 20),
    closureGroupSizeDistribution: Object.fromEntries(
      [...groupDist.entries()].sort((a, b) => a[0] - b[0]),
    ),
    qualityFlagCounts: Object.fromEntries([...flagCounts.entries()].sort((a, b) => b[1] - a[1])),
    issues,
  };

  const dir = snapshotDir(config.dataDir, snapshotId);
  await writeFile(path.join(dir, 'qa-report.json'), JSON.stringify(report, null, 2), 'utf8');
  await writeFile(path.join(dir, 'qa-summary.md'), renderMd(report), 'utf8');

  console.log(`QA report for ${snapshotId}`);
  console.log(`  links ${report.totals.links}  arcs ${report.totals.arcs}  nodes ${report.totals.nodes}`);
  console.log(`  components ${report.totals.components}, largest ${report.totals.largestComponentSharePct}%`);
  console.log(`  network length ${report.totals.totalNetworkLengthKm} km`);
  console.log(`  named links ${report.totals.namedSharePct}%`);
  console.log('\n  issues:');
  for (const i of issues) {
    console.log(`    [${i.severity.toUpperCase()}] ${i.issueType} (${i.count})`);
  }
  console.log(`\n  written: ${path.join(dir, 'qa-report.json')}`);
}

function renderMd(r: any): string {
  const L: string[] = [];
  L.push(`# QA report - ${r.snapshotId}\n`);
  L.push(`Generated ${r.generatedAtUtc}\n`);
  L.push('## Snapshot\n');
  L.push('| field | value |');
  L.push('| --- | --- |');
  for (const [k, v] of Object.entries(r.snapshot)) {
    if (k === 'notes') continue;
    L.push(`| ${k} | ${typeof v === 'object' ? `\`${JSON.stringify(v)}\`` : v} |`);
  }
  L.push('\n### Ingest notes\n');
  for (const n of r.snapshot.notes) L.push(`- ${n}`);
  L.push('\n## Totals\n');
  L.push('| metric | value |');
  L.push('| --- | --- |');
  for (const [k, v] of Object.entries(r.totals)) L.push(`| ${k} | ${v} |`);
  L.push('\n## Composition\n');
  L.push('### By model asset type\n');
  L.push('| type | links |');
  L.push('| --- | --- |');
  for (const [k, v] of Object.entries(r.byModelAssetType)) L.push(`| ${k} | ${v} |`);
  L.push('\n### By speed source\n');
  L.push('| source | links |');
  L.push('| --- | --- |');
  for (const [k, v] of Object.entries(r.bySpeedSource)) L.push(`| ${k} | ${v} |`);
  L.push('\n### Closure group size (graph links removed per physical closure)\n');
  L.push('| links in group | groups |');
  L.push('| --- | --- |');
  for (const [k, v] of Object.entries(r.closureGroupSizeDistribution)) L.push(`| ${k} | ${v} |`);
  L.push('\n### Quality flags\n');
  L.push('| flag | links |');
  L.push('| --- | --- |');
  for (const [k, v] of Object.entries(r.qualityFlagCounts)) L.push(`| ${k} | ${v} |`);
  L.push('\n## Issues\n');
  for (const i of r.issues) {
    L.push(`### [${i.severity.toUpperCase()}] ${i.issueType} - ${i.count}\n`);
    L.push(i.detail + '\n');
    if (i.sampleIds?.length) L.push('Samples: ' + i.sampleIds.map((s: string) => `\`${s}\``).join(', ') + '\n');
  }
  return L.join('\n') + '\n';
}

main().catch((e) => {
  console.error(e);
  process.exitCode = 1;
});

File: python/src/nzcl/qa.py
"""Source-data and graph quality assurance.

    nzcl-qa <snapshotId>

Writes findings to the qa_issues table and prints a summary. Nothing here
"repairs" data: the job is to measure and expose, so that a number in the
application can be traced to a stated condition of the source.
"""

from __future__ import annotations

import argparse
import json
import sys
from typing import Any

from . import db


def run(snapshot_id: str) -> dict[str, Any]:
    meta = db.query_one(
        "SELECT * FROM network_snapshots WHERE snapshot_id=%s", (snapshot_id,))
    if meta is None:
        raise SystemExit(f"unknown snapshot {snapshot_id}")

    db.execute("DELETE FROM qa_issues WHERE snapshot_id=%s", (snapshot_id,))
    issues: list[dict[str, Any]] = []

    def add(severity: str, issue_type: str, entity: str, count: int,
            detail: str, samples: list[str] | None = None) -> None:
        issues.append({"severity": severity, "issue_type": issue_type,
                       "entity_type": entity, "count": count, "detail": detail,
                       "sample_ids": samples or []})

    totals = db.query_one(
        """
        SELECT (SELECT count(*) FROM links WHERE snapshot_id=%(s)s) AS links,
               (SELECT count(*) FROM arcs WHERE snapshot_id=%(s)s) AS arcs,
               (SELECT count(*) FROM nodes WHERE snapshot_id=%(s)s) AS nodes,
               (SELECT count(DISTINCT component_id) FROM nodes WHERE snapshot_id=%(s)s) AS components,
               (SELECT round((sum(length_m)/1000.0)::numeric, 1) FROM links WHERE snapshot_id=%(s)s) AS length_km,
               (SELECT count(*) FROM links WHERE snapshot_id=%(s)s AND road_name IS NOT NULL) AS named,
               (SELECT count(*) FROM links WHERE snapshot_id=%(s)s AND oneway = 1) AS one_way,
               (SELECT count(*) FROM links WHERE snapshot_id=%(s)s AND lifeline_route) AS lifeline,
               (SELECT count(*) FROM turn_restrictions WHERE snapshot_id=%(s)s) AS restrictions,
               (SELECT count(*) FROM near_misses WHERE snapshot_id=%(s)s) AS near_misses
        """,
        {"s": snapshot_id},
    )

    # --- link-level checks -------------------------------------------------
    zero = db.query(
        "SELECT amds_id FROM links WHERE snapshot_id=%s AND length_m <= 0 LIMIT 10",
        (snapshot_id,))
    if zero:
        add("error", "ZERO_LENGTH_LINK", "link", len(zero),
            "Link has non-positive length and cannot carry traffic.",
            [r["amds_id"] for r in zero])

    loops = db.query_one(
        "SELECT count(*) AS n FROM links WHERE snapshot_id=%s "
        "AND source_node = target_node", (snapshot_id,))["n"]
    if loops:
        samples = [r["amds_id"] for r in db.query(
            "SELECT amds_id FROM links WHERE snapshot_id=%s "
            "AND source_node = target_node LIMIT 10", (snapshot_id,))]
        add("warning", "SELF_LOOP", "link", loops,
            "Link starts and ends at the same node (typically a closed loop "
            "road). Excluded from arc generation; detour results report "
            "SOURCE_DATA_ERROR.", samples)

    dupes = db.query(
        "SELECT amds_id FROM links WHERE snapshot_id=%s GROUP BY amds_id "
        "HAVING count(*) > 1 LIMIT 10", (snapshot_id,))
    if dupes:
        add("error", "DUPLICATE_STABLE_ID", "link", len(dupes),
            "Two graph links share an amds_id. Identifiers must be unique "
            "within a snapshot.", [r["amds_id"] for r in dupes])

    long_links = db.query_one(
        "SELECT count(*) AS n FROM links WHERE snapshot_id=%s AND length_m > 50000",
        (snapshot_id,))["n"]
    if long_links:
        add("warning", "IMPLAUSIBLE_LENGTH", "link", long_links,
            "Link longer than 50 km. Verify against the source geometry.")

    # --- graph structure ---------------------------------------------------
    comps = db.query(
        """
        SELECT component_id, count(*) AS links FROM (
          SELECT n.component_id, l.link_id
          FROM links l JOIN nodes n
            ON n.snapshot_id = l.snapshot_id AND n.node_id = l.source_node
          WHERE l.snapshot_id = %s
        ) q GROUP BY component_id ORDER BY links DESC
        """,
        (snapshot_id,),
    )
    total_links = totals["links"] or 1
    largest = (comps[0]["links"] / total_links) if comps else 0.0
    top_two = ((comps[0]["links"] + (comps[1]["links"] if len(comps) > 1 else 0))
               / total_links) if comps else 0.0

    # New Zealand is two main islands with no road connection between them, so
    # a national graph is legitimately dominated by TWO components. Judging it
    # on the largest alone reports Cook Strait as a defect.
    if top_two < 0.9:
        add("error", "FRAGMENTED_GRAPH", "graph", totals["components"],
            f"The two largest connected components hold only {top_two * 100:.1f}% "
            f"of links (largest {largest * 100:.1f}%). A road network should be "
            f"dominated by one component per landmass. Investigate junction "
            f"splitting.")
    else:
        add("info", "COMPONENT_STRUCTURE", "graph", totals["components"],
            f"Largest component {largest * 100:.1f}% of links; two largest "
            f"{top_two * 100:.1f}%. Two dominant components is the expected shape "
            f"for New Zealand: the North and South Islands have no road "
            f"connection. Smaller components are ferry-only islands, isolated "
            f"peninsulas, and off-network parking or access areas.")

    if totals["near_misses"]:
        add("warning", "UNCONNECTED_NEAR_MISS", "node", totals["near_misses"],
            "A link endpoint lies between the 0.05 m split tolerance and 5 m of "
            "another link but was NOT connected. Either the source has a genuine "
            "gap or the tolerance is too tight. Rows are in the near_misses table.")

    long_restrictions = db.query_one(
        "SELECT count(*) AS n FROM turn_restrictions WHERE snapshot_id=%s "
        "AND array_length(link_seq, 1) > 2", (snapshot_id,))["n"]
    add("warning" if long_restrictions else "info", "TURN_RESTRICTION_COVERAGE",
        "restriction", totals["restrictions"],
        f"{totals['restrictions']} turn restrictions are applied, of which "
        f"{long_restrictions} span more than two links. Two-link restrictions are "
        f"removed structurally from the expanded graph. NOTE: AMDS publishes only "
        f"60 restricted turns nationally, so banned-turn coverage is effectively "
        f"negligible and routes through complex intersections must not be "
        f"presented as road-legal.")

    nslr = db.query_one(
        "SELECT count(*) AS n FROM links WHERE snapshot_id=%s "
        "AND speed_source = 'nslr'", (snapshot_id,))["n"]
    add("warning" if nslr == 0 else "info", "SPEED_PROVENANCE", "link",
        total_links - nslr,
        "AMDS publishes no speed attribute. All time-metric results derive from "
        "estimated speeds and are flagged TIME_ESTIMATED. Distance is the "
        "defensible metric.")

    # --- persist -----------------------------------------------------------
    with db.direct_connection(autocommit=False) as conn:
        with conn.cursor() as cur:
            for i in issues:
                cur.execute(
                    "INSERT INTO qa_issues (snapshot_id, severity, issue_type, "
                    "entity_type, count, detail, sample_ids) "
                    "VALUES (%s,%s,%s,%s,%s,%s,%s)",
                    (snapshot_id, i["severity"], i["issue_type"], i["entity_type"],
                     i["count"], i["detail"], i["sample_ids"]),
                )
        conn.commit()

    report = {
        "snapshotId": snapshot_id,
        "totals": {
            **{k: v for k, v in totals.items()},
            "largestComponentSharePct": round(largest * 100, 2),
            "twoLargestComponentsSharePct": round(top_two * 100, 2),
        },
        "componentSizes": [c["links"] for c in comps[:20]],
        "issues": issues,
        "ingestNotes": list(meta["notes"] or []),
    }

    print(f"QA report for {snapshot_id}")
    print(f"  links {totals['links']}  arcs {totals['arcs']}  nodes {totals['nodes']}")
    print(f"  components {totals['components']}, largest {largest * 100:.2f}%, "
          f"two largest {top_two * 100:.2f}%")
    print(f"  network length {totals['length_km']} km")
    print(f"  named links {totals['named'] / total_links * 100:.2f}%")
    print("\n  issues:")
    for i in issues:
        print(f"    [{i['severity'].upper()}] {i['issue_type']} ({i['count']})")
    return report


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description="Quality report for a snapshot")
    ap.add_argument("snapshot_id", nargs="?")
    ap.add_argument("--json", action="store_true", help="print the full report as JSON")
    args = ap.parse_args(argv)

    snap = args.snapshot_id
    if not snap:
        row = db.query_one(
            "SELECT snapshot_id FROM network_snapshots ORDER BY retrieved_at_utc "
            "DESC LIMIT 1")
        if not row:
            raise SystemExit("no snapshots; run nzcl-ingest first")
        snap = row["snapshot_id"]

    report = run(snap)
    if args.json:
        print(json.dumps(report, indent=2, default=str), file=sys.stdout)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
